### ISSUE TITLE
Model forward pass outputs are identical for all inputs, but only when batch size exceeds some large value

### ISSUE THREAD

**Issue by austinhoag** [keras-team/keras#17252]
**System information**.
TensorFlow installed using miniconda, from this YAML file: https://github.com/jeffheaton/t81_558_deep_learning/blob/master/tensorflow-apple-metal-conda.yml

Platform: macOS-13.0-arm64-arm-64bit
Tensor Flow Version: 2.10.0
Keras Version: 2.10.0
Python 3.9.13

GPU model and memory: Apple M1 8 GB

**Describe the problem**.
 
I am running a simple convnet on MNIST. There are 10 output classes and the model uses a softmax activation in its output so the final outputs are probabilities for each of the 10 classes. I have found that when I do a manual forward pass (on a trained on randomly initialized model), the output is the same for each corresponding training sample. However, this only happens when the number of training samples I use in the forward pass exceeds some value (in my case around 40,000). I have looked at hundreds of other issues on here using the keyword search and none of them has addressed this problem.

**Describe the current behavior**.
Minimal working example. This is an untrained model initialized using the default initialization. I have found that the weights I set in the model don't change the behavior I am seeing. For example, if I train the model first and then evaluate it, I see this behavior.
```python
import tensorflow as tf
import tensorflow.keras

# Get the data
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Reshape and normalize the data
x_train=x_train.reshape(x_train.shape[0], x_train.shape[1], x_train.shape[2], 1)
x_train=(x_train / 255.0).astype('float32')

x_train.shape # prints (60000,28,28,1)

# Define the model and build it
num_classes = 10
input_shape = (28, 28, 1)
model = tf.keras.models.Sequential([
    tf.keras.layers.Conv2D(32, (5,5), padding='same', activation='relu', input_shape=input_shape),
    tf.keras.layers.Conv2D(32, (5,5), padding='same', activation='relu'),
    tf.keras.layers.MaxPool2D(),
    tf.keras.layers.Dropout(0.25),
    tf.keras.layers.Conv2D(64, (3,3), padding='same', activation='relu'),
    tf.keras.layers.Conv2D(64, (3,3), padding='same', activation='relu'),
    tf.keras.layers.MaxPool2D(strides=(2,2)),
    tf.keras.layers.Dropout(0.25),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.5),
    tf.keras.layers.Dense(num_classes, activation='softmax')
])
model.build(input_shape=input_shape)
```
When I run the above code I get the warning:
```python
Metal device set to: Apple M1
2022-11-17 10:03:11.366801: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:306] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
2022-11-17 10:03:11.367148: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:272] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 0 MB memory) -> physical PluggableDevice (device: 0, name: METAL, pci bus id: <undefined>)
```
No idea what `NUMA` is, but perhaps this is relevant to the issue I am seeing. 

## Do a forward pass with a small batch (size=64 training samples)
```python
batch_size=64
x_batch = x_train[0:batch_size]
x_batch.shape # prints (64,28,28,1) 

preds_batch1 = model(x_batch)
print(preds_batch1)
```
Which prints:
```python
<tf.Tensor: shape=(64, 10), dtype=float32, numpy=
array([[0.09595986, 0.1002916 , 0.10184481, 0.09941996, 0.1021289 ,
        0.10538033, 0.0985291 , 0.09900756, 0.09954101, 0.09789699],
       [0.09660684, 0.10389937, 0.10340614, 0.09841089, 0.10144418,
        0.1034528 , 0.10085803, 0.09899762, 0.09854244, 0.09438165],
       [0.09537848, 0.10107057, 0.10281242, 0.09908805, 0.10202667,
        0.10022601, 0.10605689, 0.0969203 , 0.09741242, 0.09900815],
       [0.09672208, 0.10293625, 0.10405749, 0.10432041, 0.09796295,
        0.10368358, 0.09960631, 0.09782498, 0.09770255, 0.09518337],
       [0.09558774, 0.10257296, 0.10145756, 0.10067602, 0.10090942,
        0.10331049, 0.10201143, 0.09675287, 0.09864451, 0.09807698],
       [0.09662391, 0.10230789, 0.10366919, 0.09903345, 0.09934687,
...
```
The outputs (each vector) are different for each of the 64 inputs, which is what I'd expect. No issues here.

## Do a forward pass with a big batch (size=40000 training samples)
```python
batch_size=40000
x_batch = x_train[0:batch_size]
preds_batch2 = model(x_batch)
print(preds_batch2)
```
Which prints:
```python
<tf.Tensor: shape=(40000, 10), dtype=float32, numpy=
array([[0.1, 0.1, 0.1, ..., 0.1, 0.1, 0.1],
       [0.1, 0.1, 0.1, ..., 0.1, 0.1, 0.1],
       [0.1, 0.1, 0.1, ..., 0.1, 0.1, 0.1],
       ...,
       [0.1, 0.1, 0.1, ..., 0.1, 0.1, 0.1],
       [0.1, 0.1, 0.1, ..., 0.1, 0.1, 0.1],
       [0.1, 0.1, 0.1, ..., 0.1, 0.1, 0.1]], dtype=float32)>
```
All of the output vectors are the same and they predict equal probability `0.1 = 1/10` for each class.

I have verified that the samples in `x_batch` are different. You can verify for yourself by reproducing this example.

You might ask why I would pass such a large batch in a single forward pass. I'm doing this when evaluating the model, not training. It turns out in my application I cannot use smaller batches for the evaluation step. Even if I could batch my data for evaluation, I still see this behavior as a problem. For example, how small would I need to make my batches to ensure that I won't see this unwanted behavior? But what I really want to know is: why is this happening in the first place? 

And yes, I can use a smaller test set, but I am building software that others will use. They could use a larger test set, and given that the above unwanted behavior does not raise an error (on its own), I want to understand it so I can properly safeguard against it.

**Comment by SuryanarayanaY** [comment]
HI @austinhoag ,

Thanks for reaching us. In Keras the default `kernel_initializer` is `'glorot_uniform'`. With increase in batch size the loss averages and probability will be equal to all classes as training not done on the model.This is intended and expected behaviour with `'glorot_uniform'` as `kernel_initializer`. If you observe even with low batch size the probabilities of all values nearly same (closer to 0.1 for each class).With more number of batches the total loss will be averaged and probability of each class will come closer to 0.1.

With other initializer the behaviour can be seen different.For a demo i have taken tf.keras.initializers.HeUniform() as `kernel_initializer` and the behaviour is completely different.Please refer the attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/36d9be71515badeedfaf0c3fecc5db9d/17252.ipynb#scrollTo=-uVdMZIq8pux).

**Comment by SuryanarayanaY** [comment]
Hi @austinhoag ,
I have tested the same code with both default initializer and  with other initializer `tf.keras.initializers.RandomNormal(mean=0., stddev=1.)` for same batch size of 40000 in GCP VM supported by GPU.Please find the results attached below.The code used attached [here](https://colab.research.google.com/gist/SuryanarayanaY/4d1e9ce1d0e7edaa80746fe205c99325/17252r3.ipynb).

```
(tf) suryanarayanay@ubuntu-surya-test:~$ python3 17252r3.py
2022-11-30 13:27:09.752275: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-11-30 13:27:09.881932: I tensorflow/core/util/util.cc:169] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2022-11-30 13:27:09.923354: E tensorflow/stream_executor/cuda/cuda_blas.cc:2981] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2022-11-30 13:27:10.622333: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: :/home/suryanarayanay/miniconda3/envs/tf/lib/
2022-11-30 13:27:10.622449: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: :/home/suryanarayanay/miniconda3/envs/tf/lib/
2022-11-30 13:27:10.622468: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
2022-11-30 13:27:13.789480: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-11-30 13:27:15.170668: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1616] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 38405 MB memory:  -> device: 0, name: A100-SXM4-40GB, pci bus id: 0000:00:04.0, compute capability: 8.0
2022-11-30 13:27:15.173401: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1616] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 38405 MB memory:  -> device: 1, name: A100-SXM4-40GB, pci bus id: 0000:00:05.0, compute capability: 8.0
2022-11-30 13:27:16.993250: I tensorflow/stream_executor/cuda/cuda_dnn.cc:384] Loaded cuDNN version 8100
2022-11-30 13:27:17.864793: I tensorflow/core/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2022-11-30 13:27:17.865213: I tensorflow/core/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2022-11-30 13:27:17.865246: W tensorflow/stream_executor/gpu/asm_compiler.cc:80] Couldn't get ptxas version string: INTERNAL: Couldn't invoke ptxas --version
2022-11-30 13:27:17.865618: I tensorflow/core/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2022-11-30 13:27:17.865686: W tensorflow/stream_executor/gpu/redzone_allocator.cc:314] INTERNAL: Failed to launch ptxas
Relying on driver to perform ptx compilation. 
Modify $PATH to customize ptxas location.
This message will be only logged once.
2022-11-30 13:27:20.783391: I tensorflow/stream_executor/cuda/cuda_blas.cc:1614] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.
with default kernel_initializer:glorot_uniform
tf.Tensor(
[[0.09890614 0.10303546 0.09185958 ... 0.09990875 0.09980074 0.09579183]
 [0.09710959 0.10161359 0.09540547 ... 0.09955622 0.09947632 0.09467143]
 [0.09940988 0.10129279 0.09679823 ... 0.09911065 0.09824058 0.09487872]
 ...
 [0.09701938 0.10423689 0.0952891  ... 0.09776557 0.09987197 0.09584823]
 [0.0998561  0.10280522 0.09700996 ... 0.0989674  0.09767013 0.09529291]
 [0.09885789 0.10391052 0.09469009 ... 0.09886216 0.09852277 0.09453542]], shape=(40000, 10), dtype=float32)
/home/suryanarayanay/miniconda3/envs/tf/lib/python3.9/site-packages/keras/initializers/initializers_v2.py:120: UserWarning: The initializer RandomNormal is unseeded and being called multiple times, which will return identical values  each time (even if the initializer is unseeded). Please update your code to provide a seed to the initializer, or avoid using the same initalizer instance more than once.
  warnings.warn(
with  other kernel_initializer:RandomNormal()
tf.Tensor(
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 1. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]], shape=(40000, 10), dtype=float32)
(tf) suryanarayanay@ubuntu-surya-test:~$ 
```
As you can see there is clear difference in predictions because of different intilzers we used.The precision of outputs might also varies from CPU to GPU and also GPU to GPU.
I hope you also have clear understanding of Normal and Uniform distributions which is making difference in predictions above. You may also refer some resources from web here [1](https://www.analyticsvidhya.com/blog/2017/09/6-probability-distributions-data-science/),  [2](https://datasciencedojo.com/blog/types-of-statistical-distributions-in-ml/) incase you want to check.

I hope this will clarify your query.

Thankyou!

**Comment by austinhoag** [comment]
Hi, 

Thanks for the response. I understand what normal and uniform distributions are, and I understand that by using those distributions to set the weights of a model, one would get different predictions for the same inputs. However, I don't understand how the output you showed demonstrates my issue. Your outputs from the glorot_uniform kernel initializer are not the same for each sample, which is what I was seeing for a batch size of 40,000. I see that your outputs are all _close_ to 0.1 for each class in each sample, but not exactly 0.1. 

I guess my underlying question is: why does batch size affect predictions. You stated this an explanation in your initial response in this thread:

> With increase in batch size the loss averages and probability will be equal to all classes as training not done on the model

Can you explain this further? I don't understand this. As far as I am aware, there is no loss being calculated here. All we are doing is setting the weights of the model and then passing some inputs through a single forward pass of the model.

**Comment by SuryanarayanaY** [comment]
Hi @austinhoag ,
> I see that your outputs are all _close_ to 0.1 for each class in each sample, but not exactly 0.1.

The results will vary slightly from GPU to GPU due to their own internal configurations and default precisions etc.Some GPUs might be rounding off the values too.Hence we can't exactly compare values of different GPUs to exact precision.

Can you can please try changing the dtype of `x_train` to `float64` then there might be change in the output values too w.r.t precision.Please try this and let us know the results.

Thankyou!

**Comment by SuryanarayanaY** [comment]
Hi @austinhoag ,

As i mentioned earlier the precision changes from GPU to GPU also.For example the results that I have attached in above [comment](https://github.com/keras-team/keras/issues/17252#issuecomment-1332188465) is from a GCP VM with Ubuntu 18.04 OS and with NVIDIA A100 40GB GPU. Request you to please check the results with different GPUs to observe the behaviour.

Thankyou!

**Comment by austinhoag** [comment]
I don't have access to a different GPU at the moment, but based on your results it does seem like the GPU being used can change the results. This is really the only somewhat satisfying answer to my original issue:

> Passing a different number of inputs through a (untrained) model's forward pass can change the model's output for the same subset of inputs. There seems to be a threshold number at which results start to differ, which I would imagine is model-dependent.

However, given that you have not been able to reproduce my issue on your end, I'm still not 100% sure this is what is causing this. For anyone who does get the same issue, the solution I found was to keep batch sizes relatively small. It's unfortunate that there is no warning when you reach the sufficiently large batch size to cause this issue. So you need to be quite careful.
